File: restaurant_system/frontend/lib/core/models/order.dart
class OrderItem {
  final int menuItemId;
  final int quantity;
  final String? specialInstructions;
  final String? menuItemName;
  final double? menuItemPrice;

  OrderItem({
    required this.menuItemId,
    required this.quantity,
    this.specialInstructions,
    this.menuItemName,
    this.menuItemPrice,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      menuItemId: json['menu_item_id'] ?? 0,
      quantity: json['quantity'] ?? 0,
      specialInstructions: json['special_instructions'],
      menuItemName: json['menu_items']?['name'],
      menuItemPrice: json['menu_items']?['price']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'menu_item_id': menuItemId,
      'quantity': quantity,
      'special_instructions': specialInstructions,
    };
  }

  OrderItem copyWith({
    int? menuItemId,
    int? quantity,
    String? specialInstructions,
    String? menuItemName,
    double? menuItemPrice,
  }) {
    return OrderItem(
      menuItemId: menuItemId ?? this.menuItemId,
      quantity: quantity ?? this.quantity,
      specialInstructions: specialInstructions ?? this.specialInstructions,
      menuItemName: menuItemName ?? this.menuItemName,
      menuItemPrice: menuItemPrice ?? this.menuItemPrice,
    );
  }

  double get totalPrice => (menuItemPrice ?? 0) * quantity;
}

class Order {
  final int? id;
  final int tableNumber;
  final List<OrderItem> items;
  final String status;
  final double totalAmount;
  final String? customerName;
  final String? specialInstructions;
  final DateTime? createdAt;
  final String? paymentStatus;

  Order({
    this.id,
    required this.tableNumber,
    required this.items,
    this.status = 'pending',
    required this.totalAmount,
    this.customerName,
    this.specialInstructions,
    this.createdAt,
    this.paymentStatus,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    List<OrderItem> orderItems = [];
    
    if (json['order_items'] != null) {
      orderItems = (json['order_items'] as List)
          .map((item) => OrderItem.fromJson(item))
          .toList();
    } else if (json['items'] != null) {
      orderItems = (json['items'] as List)
          .map((item) => OrderItem.fromJson(item))
          .toList();
    }

    return Order(
      id: json['id'],
      tableNumber: json['table_number'] ?? 0,
      items: orderItems,
      status: json['status'] ?? 'pending',
      totalAmount: (json['total_amount'] ?? 0).toDouble(),
      customerName: json['customer_name'],
      specialInstructions: json['special_instructions'],
      createdAt: json['created_at'] != null 
          ? DateTime.tryParse(json['created_at']) 
          : null,
      paymentStatus: json['payment_status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'table_number': tableNumber,
      'items': items.map((item) => item.toJson()).toList(),
      'status': status,
      'total_amount': totalAmount,
      'customer_name': customerName,
      'special_instructions': specialInstructions,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      'payment_status': paymentStatus,
    };
  }

  Order copyWith({
    int? id,
    int? tableNumber,
    List<OrderItem>? items,
    String? status,
    double? totalAmount,
    String? customerName,
    String? specialInstructions,
    DateTime? createdAt,
    String? paymentStatus,
  }) {
    return Order(
      id: id ?? this.id,
      tableNumber: tableNumber ?? this.tableNumber,
      items: items ?? this.items,
      status: status ?? this.status,
      totalAmount: totalAmount ?? this.totalAmount,
      customerName: customerName ?? this.customerName,
      specialInstructions: specialInstructions ?? this.specialInstructions,
      createdAt: createdAt ?? this.createdAt,
      paymentStatus: paymentStatus ?? this.paymentStatus,
    );
  }

  String get statusDisplayName {
    switch (status.toLowerCase()) {
      case 'pending':
        return 'Pending';
      case 'confirmed':
        return 'Confirmed';
      case 'preparing':
        return 'Preparing';
      case 'ready':
        return 'Ready';
      case 'served':
        return 'Served';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status;
    }
  }

  bool get isPaid => paymentStatus == 'paid';
  bool get isCompleted => status == 'served' || status == 'cancelled';
}

File: backend/models/order.py
from sqlalchemy import Column, Integer, String, DateTime, Boolean, Text, ForeignKey, Float, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from database import Base

class Order(Base):
    __tablename__ = "orders"

    id = Column(Integer, primary_key=True, index=True)
    restaurant_id = Column(Integer, ForeignKey("restaurants.id"), nullable=False)
    table_id = Column(Integer, ForeignKey("tables.id"), nullable=False)
    customer_id = Column(Integer, ForeignKey("users.id"), nullable=True)  # Nullable for guest orders
    
    # Order details
    order_number = Column(String(50), unique=True, nullable=False, index=True)
    status = Column(String(50), default="pending")  # pending, confirmed, preparing, ready, served, cancelled
    
    # Pricing
    subtotal = Column(Float, nullable=False)
    tax_amount = Column(Float, default=0.0)
    service_charge = Column(Float, default=0.0)
    discount_amount = Column(Float, default=0.0)
    total_amount = Column(Float, nullable=False)
    
    # Payment
    payment_method = Column(String(50), nullable=True)  # mpesa, card, cash
    payment_status = Column(String(50), default="pending")  # pending, paid, failed, refunded
    
    # Customer information (for guest orders)
    customer_name = Column(String(255), nullable=True)
    customer_phone = Column(String(20), nullable=True)
    customer_email = Column(String(255), nullable=True)
    
    # Special instructions
    special_instructions = Column(Text, nullable=True)
    
    # Timing
    estimated_preparation_time = Column(Integer, nullable=True)  # minutes
    preparation_started_at = Column(DateTime(timezone=True), nullable=True)
    ready_at = Column(DateTime(timezone=True), nullable=True)
    served_at = Column(DateTime(timezone=True), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    restaurant = relationship("Restaurant", back_populates="orders")
    table = relationship("Table", back_populates="orders")
    customer = relationship("User")
    order_items = relationship("OrderItem", back_populates="order", cascade="all, delete-orphan")
    payments = relationship("Payment", back_populates="order")
    
    def __repr__(self):
        return f"<Order(id={self.id}, number='{self.order_number}', status='{self.status}')>"

class OrderItem(Base):
    __tablename__ = "order_items"

    id = Column(Integer, primary_key=True, index=True)
    order_id = Column(Integer, ForeignKey("orders.id"), nullable=False)
    menu_item_id = Column(Integer, ForeignKey("menu_items.id"), nullable=False)
    
    # Item details
    quantity = Column(Integer, nullable=False, default=1)
    unit_price = Column(Float, nullable=False)
    total_price = Column(Float, nullable=False)
    
    # Customizations
    customizations = Column(JSON, nullable=True)  # JSON object with customization choices
    special_instructions = Column(Text, nullable=True)
    
    # Status
    status = Column(String(50), default="pending")  # pending, preparing, ready, served
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    order = relationship("Order", back_populates="order_items")
    menu_item = relationship("MenuItem", back_populates="order_items")
    
    def __repr__(self):
        return f"<OrderItem(id={self.id}, quantity={self.quantity}, price={self.total_price})>"

class ServiceRequest(Base):
    __tablename__ = "service_requests"

    id = Column(Integer, primary_key=True, index=True)
    table_id = Column(Integer, ForeignKey("tables.id"), nullable=False)
    customer_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    
    # Request details
    request_type = Column(String(50), nullable=False)  # water, bill, assistance, cleanup
    message = Column(Text, nullable=True)
    priority = Column(Integer, default=1)  # 1=low, 2=medium, 3=high, 4=urgent
    
    # Status
    status = Column(String(50), default="pending")  # pending, acknowledged, in_progress, completed, cancelled
    
    # Assignment
    assigned_to = Column(Integer, ForeignKey("users.id"), nullable=True)  # Staff member assigned
    
    # Timing
    acknowledged_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    table = relationship("Table", back_populates="service_requests")
    customer = relationship("User", foreign_keys=[customer_id])
    assigned_staff = relationship("User", foreign_keys=[assigned_to])
    
    def __repr__(self):
        return f"<ServiceRequest(id={self.id}, type='{self.request_type}', status='{self.status}')>"

